### ISSUE TITLE
Fine-tunning of pre-trained MiniLM

### ISSUE THREAD

**Issue by deepaknlp** [microsoft/unilm#147]
Hi @donglixp,

Is it possible to use pre-trained MiniLM to initialize the seq2seq fine tunning? I tried but got very bad results. I wonder why the fine-tuning using masked word prediction giving good results compare to standard seq2seq fine-tunning.

Looking for help!

**Comment by nyanyanya** [comment]
Hi @deepaknlp,

What MiniLM model did you use in s2s fine-tuning? Currently the s2s-ft package does not work with multilingual MiniLM model.

Thanks,
Nan

**Comment by deepaknlp** [comment]
Thank you @nyanyanya  for the quick response.
I am using [this](https://1drv.ms/u/s!AjHn0yEmKG8qixAYyu2Fvq5ulnU7?e=DFApTA) MiniLM model. I implemented my own encoder-decoder based model and want to initialize the parameters with MiniLM for generation task. 

However, the performance decrease significantly compared to the masked word prediction objective as implemented in s2s-ft package.  I am looking for a technical suggestion that whether the MiniLM is suitable to fine-tune on encoder-decoder based generation task, I believe so because MiniLM is built on  the self-attention+distillation architecture on BERT.

Let me know if I am missing something here.

**Comment by donglixp** [comment]
The recommended recipe for NLG fine-tuning is using the s2s-ft package as in https://github.com/microsoft/unilm/tree/master/minilm#fine-tuning-on-nlg-tasks . More details about how you used MiniLM to initialize the encoder and decoder would be helpful to solve the problem.

**Comment by deepaknlp** [comment]
Hey @donglixp,

I understood the recommended procedure. In my experiment, I am using the [BERT implementation](https://github.com/microsoft/unilm/blob/4d5c1cd07b43b3564c03aa1722a76a56aeadbc68/s2s-ft/s2s_ft/modeling.py#L346) of the s2s-ft package as an encoder and [this](https://github.com/huggingface/transformers/blob/a34a9896ac2a4a33ff9cd805c76eed914c8d8965/src/transformers/modeling_bert.py#L594) BERT model as a decoder. I initialized the model parameters with MiniLM weights. However, the encoder-decoder attention is initialized with the random weights.

**Comment by WenhuiWang0824** [comment]
Hi @deepaknlp,

You use MiniLM to both initialize the encoder and decoder weights? I am reading the codes you provided and will try to answer the question later.

**Comment by deepaknlp** [comment]
Thank you @WenhuiWang0824 and @donglixp.
It is much appreciated. Yes, I used MiniLM to both initialize the encoder and decoder weights.

**Comment by WenhuiWang0824** [comment]
Hi @deepaknlp,

For the decoder code, the representation of the current token is used to predict the next token (traditional LM, such as GPT)? If so, I think you could refer to the implementation in s2s-ft, which uses a mask token to predict the current token. Using the mask token to predict tokens may perform better given the training objective in the teacher model.

Besides, parameters of the encoder-decoder attention are randomly initialized. The resulting cross attention representation may affect the original attention and hidden outputs in MiniLM. In the s2s-ft package, we do not use encoder-decoder attention and fine-tune the original MINILM as a sequence-to-sequence model by employing a specific self-attention mask. The introduced encoder-decoder attention parameters may also affect the results.

Thanks

**Comment by deepaknlp** [comment]
Thank you @WenhuiWang0824 for the detailed insights.
Yes, I was using the current token to predict the next token. I agree with you to use the mask token to predict the next token, It makes sense. The encoder-decoder attention parameters may be the problem here as you rightly pointed.

Thank you for your time.
